cluster/addons.go

```diff
@@ -2,32 +2,132 @@ package cluster
 
 import (
 	"fmt"
+	"time"
 
+	"github.com/rancher/rke/addons"
+	"github.com/rancher/rke/k8s"
 	"github.com/sirupsen/logrus"
 )
 
 const (
-	ClusterDNSServerIPEnvName = "RKE_DNS_SERVER"
-	ClusterDomainEnvName      = "RKE_CLUSTER_DOMAIN"
+	KubeDNSAddonResourceName = "rke-kubedns-addon"
+	UserAddonResourceName    = "rke-user-addon"
 )
 
 func (c *Cluster) DeployK8sAddOns() error {
 	err := c.deployKubeDNS()
 	return err
 }
 
+func (c *Cluster) DeployUserAddOns() error {
+	logrus.Infof("[addons] Setting up user addons..")
+	if c.Addons == "" {
+		logrus.Infof("[addons] No user addons configured..")
+		return nil
+	}
+
+	if err := c.doAddonDeploy(c.Addons, UserAddonResourceName); err != nil {
+		return err
+	}
+	logrus.Infof("[addons] User addon deployed successfully..")
+	return nil
+
+}
+
 func (c *Cluster) deployKubeDNS() error {
-	logrus.Infof("[plugins] Setting up KubeDNS")
+	logrus.Infof("[addons] Setting up KubeDNS")
+
+	kubeDNSYaml := addons.GetKubeDNSManifest(c.ClusterDNSServer, c.ClusterDomain)
+
+	if err := c.doAddonDeploy(kubeDNSYaml, KubeDNSAddonResourceName); err != nil {
+		return err
+	}
+	logrus.Infof("[addons] KubeDNS deployed successfully..")
+	return nil
+
+}
+
+func (c *Cluster) doAddonDeploy(addonYaml, resourceName string) error {
 
-	kubectlCmd := &KubectlCommand{
-		Cmd: []string{"apply -f /addons/kubedns*.yaml"},
+	err := c.StoreAddonConfigMap(addonYaml, resourceName)
+	if err != nil {
+		return fmt.Errorf("Failed to save addon ConfigMap: %v", err)
 	}
-	logrus.Infof("[plugins] Executing the deploy command..")
-	err := c.RunKubectlCmd(kubectlCmd)
+
+	logrus.Infof("[addons] Executing deploy job..")
+
+	addonJob := c.GetAddonsExcuteJob(resourceName)
+	err = c.ApplySystemAddonExcuteJob(addonJob)
+	if err != nil {
+		return fmt.Errorf("Failed to deploy addon execute job: %v", err)
+	}
+	return nil
+}
+
+func (c *Cluster) StoreAddonConfigMap(addonYaml string, addonName string) error {
+	logrus.Infof("[addons] Saving addon ConfigMap to Kubernetes")
+	kubeClient, err := k8s.NewClient(c.LocalKubeConfigPath)
 	if err != nil {
-		return fmt.Errorf("Failed to run kubectl command: %v", err)
+		return err
+	}
+	timeout := make(chan bool, 1)
+	go func() {
+		for {
+			err := k8s.UpdateConfigMap(kubeClient, []byte(addonYaml), addonName)
+			if err != nil {
+				time.Sleep(time.Second * 5)
+				fmt.Println(err)
+				continue
+			}
+			logrus.Infof("[addons] Successfully Saved addon to	 Kubernetes ConfigMap: %s", addonName)
```

Comment:
I think this is an editor problem that added an extra tab to log info, can you remove that